DM-2433 Operations clean-up after agent crash. (#1175)
SoftwareAG/c8y-docs: content/reference/device-management-library-bundle/general-concepts.md

**SmartREST 2.0**

> {{< product-c8y-iot >}} provides the static templates 501, 502, and 503 to manipulate the operation status. These templates take the operation type as input parameter and always update the oldest operation in the preceding status. We recommend handling all operations sequentially in the order they arrive at the device 

Reviewer: ```suggestion
{{< product-c8y-iot >}} provides the static templates 501, 502, and 503 to manipulate the operation status. These templates take the operation type as input parameter and always update the oldest operation in the preceding status. We recommend you to let all operations be handled sequentially in the order they arrive at the device.
```

> Devices can find their operations IDs by querying the Device Control API, by subscribing to the operation JSON topic, or by using a custom response template which includes the IDs, that is, template 504, 505, or 506, which enable setting the status of operations with a known ID.

Reviewer: ```suggestion
Devices can find their operation IDs by querying the Device Control API, by subscribing to the operation JSON topic, or by using a custom response template which includes the IDs, that is, template 504, 505, or 506, which enable setting the status of operations with a known ID.
```

### Error handling during operation processing

If any error occurs during the processing of an operation the device must set the operation status to FAILED and provide a failure reason as descriptive as possible. This includes any unexpected or expected error conditions that prevent the operation to be fully completed and as expected. Even if only one step in an operation with multiple distinct steps fails, the entire operation must be considered as FAILED.

It is up to the device and its use case whether it should roll back any local state changes that happened before the error occurred. If any change of state remains after an operation failed the device must communicate this changed state with {{< product-c8y-iot >}}.

### Recovering after agent crash

> After unexpected restart device must set status to FAILED for operations which processing was interrupted. First it needs to learn the operations ids:

Reviewer: ```suggestion
After an unexpected restart a device must set the status to FAILED for operations for which processing was interrupted. First it must learn the operation IDs:
```

```http
GET /devicecontrol/operations?deviceId=<deviceId>&status=EXECUTING
```

```json
{
  "operations": [
    {
      "creationTime": "2023-06-25T14:53:52.395Z",
      "deviceId": "123",
      "id": "101",
      "status": "EXECUTING",
      "c8y_Restart": {}
    },
    {
      "creationTime": "2023-06-25T14:57:29.089Z",
      "deviceId": "123",
      "id": "102",
      "status": "EXECUTING",
      "c8y_SendConfiguration": {}
    }
  ]
}
```

> Then one by one change the statuses:

Reviewer: ```suggestion
Then it must change the statuses one by one:
```

```http
PUT /devicecontrol/operations/<operationId>
```

```json
{
  "status": "FAILED"
}
```

**SmartREST 2.0**

> Alternatively the static template 507 may be used. The template changes status from EXECUTING to FAILED for all operations of given type or of any type.

Reviewer: ```suggestion
Alternatively the static template 507 may be used. The template changes the status from EXECUTING to FAILED for all operations of the given type or for all types.
```

Devices can find their operation IDs by querying the Device Control API, by subscribing to the operation JSON topic, or by using a custom response template which includes the IDs, that is, template 504, 505, or 506, which enable setting the status of operations with a known ID.

### Error handling during operation processing

If any error occurs during the processing of an operation the device must set the operation status to FAILED and provide a failure reason as descriptive as possible. This includes any unexpected or expected error conditions that prevent the operation to be fully completed and as expected. Even if only one step in an operation with multiple distinct steps fails, the entire operation must be considered as FAILED.

It is up to the device and its use case whether it should roll back any local state changes that happened before the error occurred. If any change of state remains after an operation failed the device must communicate this changed state with {{< product-c8y-iot >}}.

### Recovering after agent crash

> After an unexpected restart a device must cleanly recover its status. This includes all status parameters communicated with the platform and all ongoing operations. Recovering status can be done by updating all values in the cloud with the current truth from the device. Recovering ongoing operations is more difficult. Devices are expected to keep track of all operations they moved to status EXECUTING. Typically devices keep information of longer running operations in a persistent storage so that they can be resumed. In unexpected shutdowns or crash scenarios this may not always be possible. In this case the device may cancel all ongoing operations to reset its own status.

Reviewer: ```suggestion
After an unexpected restart a device must cleanly recover its status. This includes all status parameters communicated with the platform and all ongoing operations. Recovering the status can be done by updating all values in the cloud with the current values on the device. Recovering ongoing operations is more difficult. Devices are expected to keep track of all operations they moved to status EXECUTING. Typically devices keep information of longer-running operations in a persistent storage so that they can be resumed. In unexpected shutdown or crash scenarios this may not always be possible. In this case the device may cancel all ongoing operations to reset its own status.
```

{{< product-c8y-iot >}} provides the static templates 501, 502, and 503 to manipulate the operation status. These templates take the operation type as input parameter and always update the oldest operation in the preceding status. We recommend handling all operations sequentially in the order they arrive at the device 

> Device may get to know its operations ids by querying devicecontrol API, by subscribing to operation JSON topic, or by using custom response template which includes the ids. In these cases device may choose to use templates 504, 505, 506, which enables setting status of operations with known id.

Reviewer 2: ```suggestion
Devices can find their operations IDs by querying the Device Control API, by subscribing to the operation JSON topic, or by using a custom response template which includes the IDs, that is, template 504, 505, or 506, which enable setting the status of operations with a known ID.
```
